File: pysrc/Integ.cpp
/* -*- c++ -*-
 * Copyright (c) 2012-2021 by the GalSim developers team on GitHub
 * https://github.com/GalSim-developers
 *
 * This file is part of GalSim: The modular galaxy image simulation toolkit.
 * https://github.com/GalSim-developers/GalSim
 *
 * GalSim is free software: redistribution and use in source and binary forms,
 * with or without modification, are permitted provided that the following
 * conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions, and the disclaimer given in the accompanying LICENSE
 *    file.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions, and the disclaimer given in the documentation
 *    and/or other materials provided with the distribution.
 */

#include "PyBind11Helper.h"
#include "integ/Int.h"
#include "math/Hankel.h"
#include <iostream>

namespace galsim {
namespace integ {

#if defined(__GNUC__) && __GNUC__ >= 6
// Workaround for a bug in some versions of gcc 6-8.
// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80947
#pragma GCC visibility push(hidden)
#endif
    // A C++ function object that just calls a python function.
    class PyFunc :
        public std::unary_function<double, double>
    {
    public:
        PyFunc(const py::function& func) : _func(func) {}
        double operator()(double x) const
        { return PY_CAST<double>(_func(x)); }
    private:
        const py::function& _func;
    };
#if defined(__GNUC__) && __GNUC__ >= 6
#pragma GCC visibility pop
#endif

    // Integrate a python function using int1d.
    py::tuple PyInt1d(const py::function& func, double min, double max,
                      double rel_err=DEFRELERR, double abs_err=DEFABSERR)
    {
        PyFunc pyfunc(func);
        try {
            double res = int1d(pyfunc, min, max, rel_err, abs_err);
            return py::make_tuple(true, res);
        } catch (IntFailure& e) {
            return py::make_tuple(false, e.what());
        }
    }

    // Integrate a python function using int1d.
    void PyHankel(const py::function& func, size_t ik, size_t ires, int N,
                  double nu, double rmax,
                  double rel_err=DEFRELERR, double abs_err=DEFABSERR)
    {
        const double* k = reinterpret_cast<const double*>(ik);
        double* res = reinterpret_cast<double*>(ires);
        PyFunc pyfunc(func);
        if (rmax == 0.) {
            for (int i=0; i<N; ++i) {
                res[i] = math::hankel_inf(pyfunc, k[i], nu, rel_err, abs_err);
            }
        } else {
            for (int i=0; i<N; ++i) {
                res[i] = math::hankel_trunc(pyfunc, k[i], nu, rmax, rel_err, abs_err);
            }
        }
    }

    void pyExportInteg(PY_MODULE& _galsim)
    {
        GALSIM_DOT def("PyInt1d", &PyInt1d);
        GALSIM_DOT def("PyHankel", &PyHankel);
    }

} // namespace integ
} // namespace galsim

File: galsim/deprecated/integ.py
# Copyright (c) 2012-2021 by the GalSim developers team on GitHub
# https://github.com/GalSim-developers
#
# This file is part of GalSim: The modular galaxy image simulation toolkit.
# https://github.com/GalSim-developers/GalSim
#
# GalSim is free software: redistribution and use in source and binary forms,
# with or without modification, are permitted provided that the following
# conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions, and the disclaimer given in the accompanying LICENSE
#    file.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions, and the disclaimer given in the documentation
#    and/or other materials provided with the distribution.
#

import numpy as np
from functools import reduce
import galsim
from ..errors import GalSimRangeError

def trapz(func, min, max, points=10000):
    """Simple wrapper around 'numpy.trapz' to take function and limits as inputs.

    Example usage::

        >>> def func(x): return x**2
        >>> galsim.integ.trapz(func, 0, 1)
        0.33333333500033341
        >>> galsim.integ.trapz(func, 0, 1, 1e6)
        0.33333333333349996
        >>> galsim.integ.trapz(func, 0, 1, np.linspace(0, 1, 1e3))
        0.33333350033383402

    Parameters:
        func:       The function to be integrated.  y = func(x) should be valid.
        min:        The lower end of the integration bounds.
        max:        The upper end of the integration bounds.
        points:     If integer, the number of points to sample the integrand. If array-like, then
                    the points to sample the integrand at. [default: 1000].
    """
    from . import depr
    depr('galsim.integ.trapz', 2.3, 'galsim.integ.int1d')

    if not np.isscalar(points):
        if (np.max(points) > max) or (np.min(points) < min):
            raise GalSimRangeError("Points outside of specified range", points, min, max)
    elif int(points) != points:
        raise TypeError("npoints must be integer type or array")
    else:
        points = np.linspace(min, max, points)

    return np.trapz(func(points),points)

def midpt(fvals, x):
    """Midpoint rule for integration.

    Parameters:
        fvals:    Samples of the integrand
        x:        Locations at which the integrand was sampled.

    Returns:
        midpoint rule approximation of the integral.
    """
    from . import depr
    depr('galsim.integ.midpt', 2.3, 'np.trapz or galsim.trapz')

    dx = [x[1]-x[0]]
    dx.extend(0.5*(x[2:]-x[0:-2]))
    dx.append(x[-1]-x[-2])
    weighted_fvals = [w*f for w,f in zip(dx, fvals)]
    return reduce(lambda y,z:y+z, weighted_fvals)


galsim.integ.trapz = trapz
galsim.integ.midpt = midpt
